### ISSUE TITLE
Problem with "Conflicted Copy" notes + tags and note history

### ISSUE THREAD

**Issue by bschlagel** [standardnotes/forum#678]
Just experienced a weird/disconcerting issue with *five* copies of a note being created with the "Conflicted Copy" badge.

This is a note that I was editing on my iPhone while traveling. It seemed to be syncing fine via mobile data, though possible I closed the app before it finished syncing or something. When I got home I opened the app on my phone again and noticed all the duplicate "Conflicted Copy" notes. I opened the Mac desktop app and saw the same thing. The actual "Conflicted Copy" badge disappeared after I opened the notes, but all five duplicates are still in the notes list.

The note in question wasn't edited in a way where there should be unresolvable diffs. It's simply a list that I added more items to. Nothing deleted or edited. So there are now five copies, with the newer ones just having more text added at the top of the note. (A bit confusing in the first place as I'd expect all these dupes could be merged automatically.)

Another odd thing here: not all the copies have tags. The oldest one has the correct tags. The next one has *no* tags. The two after that have the correct tags. The newest one (only a single line of text different from the previous one) has no tags. So, there may be a bug here related to tags?

Similarly, could be an issue related to "Note History Action Menu", noticed something unexpected here as well. The *second newest* note is the one with the full "note history", i.e. seems this is the "real" or "original" note. The newest one in the list, and the three oldest ones, have just one or two items in the Note History Action Menu. Those four all have an identical timestamp in the newest item in the NHAM; the second newest has a timestamp one minute earlier.

Anyway, seems like possibly some combination of issues. For one, not sure how so many copies got made; I'd expect two e.g. new unsynced one on phone vs. older copy, but not five. And then, not sure if separate or the source of the dupes in the first place, the inconsistencies with tags and note history.

I didn't edit / delete any of the dupes yet in case it may be helpful to get you any further info on this. Let me know.

**Comment by mobitar** [comment]
Sync conflicts in general don't keep count. If there is a conflict, a duplicate will unconditionally be made. Sync conflicts occur if the date of the item you are trying to save is different from the date the server has. In this case, it's likely that there were some dropped requests when you were saving from mobile, causing a date mismatch. And indeed, only the original note will have the full note history, as history is tracked by note id, and duplicates get a new id. This issue isn't really disconcerting, but it is inconvenient. We'll definitely want to improve on this in some way in the future.

**Comment by bschlagel** [comment]
Ah okay thanks for clarifying! Yeah I guess I was on a shaky mobile connection and a few duplicates got queued up behind the scenes but I only noticed when I got home an opened the app again.

Maybe still something to look into here re: tags — not sure why the original (note w/ full history) and two of the dupes have the correct tags, but the other two dupes ended up with no tags at all.

**Comment by bschlagel** [comment]
Following up as I continue to run into this issue occasionally and just had to again delete a half dozen duplicates of a single note.

I see why duplicate notes are created in the instance of a conflict. But to improve here, would there potentially be a way to auto-merge dupe notes without the possibility of data loss?

This seems to happen when a save fails due to being offline, but couldn't the dupes be detected and automatically merged (with most recent copy taking precedence) upon reconnecting? Or is the issue that due to sync errors it can sometimes be impossible to tell what version of a note is actually newest, or some similar source of unresolvable merge conflicts?

**Comment by mobitar** [comment]
Merging is non-trivial, especially client side. But if you've found upon a flow that consistently results in excessive duplicates, let me know. I've yet to replicate any such scenarios.

**Comment by bschlagel** [comment]
Gotcha. My specific use case: I have a 'book list' note where I jot down titles of interesting books whenever I'm browsing in a bookstore, using the iOS app. 

It doesn't usually cause duplicates, but has done so several times. I'm sure being offline / spotty connection has something to do with it. Maybe also opening / closing the app often has something to do with it as well? That is: I'll open the app briefly, write down a book title (just a few words / few seconds) then lock the phone and put in my pocket. Open again a couple minutes later and repeat. Might do that many times in say an hour; some of the times I add things may be w/o solid mobile connection. Then later I'll be checking the notes list and realize a handful of duplicates have gotten created.

**Comment by mobitar** [comment]
During this flow, do you typically wait for the app to go from saying "Saving..." to "All changes saved", or do you just exit immediately without waiting for that confirmation?

**Comment by bschlagel** [comment]
It varies, I sometimes notice the 'all changes saved' but no I'm not always paying attention to that. I think I do sometimes just hit the lock button so I can put my phone back in my pocket as soon as I'm done typing.

**Comment by zedgoat** [comment]
Adding experience report here. I wonder if **pinning** the Conflicted Copy badge with a manual confirmation (instead of automatically clearing the status on select) would help with the UX?

---

Preface: I recovered the content from a backup so there was no actual data loss here.

- In evening, workstation went offline while editing note, SN didn't reconnect (Linux machine w/ VPN that screws up DNS).
- Laptop next morning: conflicted note appears, both were similar (using Meld for compare) apart from 1 paragraph, pasted in merge and permanently deleted one copy.
- Editing on laptop, noticed missing content block earlier in the doc (it's a rolling logbook). Returned to workstation, content was there and SN was flagging that sync wasn't connected.
- Cycled VPN connection, SN synced an note with missing content was "replaced" with new merged note, losing content. Looking at Note History Action Menu, this is the Conflict note that has synced as it only had today's history.
- Recovered content from local SN backup directory.

Suspect the conflict on the Laptop came from a third source, then I've merged, tabbed back into SN, populated the **conflict** copy and then deleted the **original** (instead of the other way around, as the Conflicted badge had cleared). Then, when connectivity on Workstation was resolved, the delete action synced for the original and I've been left with the "new" conflict note.

Putting this down to user error at several steps here, but the workflow could catch people out?
